contracts/YieldMath.sol

```diff
@@ -334,6 +334,29 @@ library YieldMath {
   uint256 public constant MAX = type(uint128).max;   // Used for overflow checks
   uint256 public constant VAR = 1e12;                // The logarithm math used is not precise to the wei, but can deviate up to 1e12 from the real value.
 
+  /**
+   * Calculate a YieldSpace pool invariant according to the whitepaper
+   */
+  function invariant(uint128 baseReserves, uint128 fyTokenReserves, uint128 timeTillMaturity, int128 k)
+      public pure returns(uint128)
+  {
+    unchecked {
+      // a = (1 - k * timeTillMaturity)
+      int128 a = int128(ONE).sub(k.mul(timeTillMaturity.fromUInt()));
+      require (a > 0, "YieldMath: Too far from maturity");
+
+      uint256 sum =
+      uint256(baseReserves.pow(uint128 (a), ONE)) +
+      uint256(fyTokenReserves.pow(uint128 (a), ONE)) >> 1;
+      require(sum < MAX, "YieldMath: Sum overflow");
+
+      uint256 result = uint256(uint128(sum).pow(ONE, uint128(a))) << 1;

```

Comment:
Instead of `<< 1`, it needs to be divided by `S` (the totalSupply).

Include a function in each Pool as well to calculate this.